### PR TITLE
[SPARK-38675][CORE] Fix race during unlock in BlockInfoManager

### PR DESCRIPTION
**Pull request by hvanhovell**
### What changes were proposed in this pull request?
This PR fixes a race in the `BlockInfoManager` between `unlock` and `releaseAllLocksForTask`, resulting in a negative reader count for a block (which trips an assert). This happens when the following events take place:

1. [THREAD 1] calls `releaseAllLocksForTask`. This starts by collecting all the blocks to be unlocked for this task.
2. [THREAD 2] calls `unlock` for a read lock for the same task (this means the block is also in the list collected in step 1). It then proceeds to unlock the block by decrementing the reader count.
3. [THREAD 1] now starts to release the collected locks, it does this by decrementing the readers counts for blocks by the number of acquired read locks. The problem is that step 2 made the lock counts for blocks incorrect, and we decrement by one (or a few) too many. This triggers a negative reader count assert.

We fix this by adding a check to `unlock` that makes sure we are not in the process of unlocking. We do this by checking if there is a multiset associated with the task that contains the read locks.

### Why are the changes needed?
It is a bug. Not fixing this can cause negative reader counts for blocks, and this causes task failures.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added a regression test in BlockInfoManager suite.

### CODE REVIEW

**Comment by hvanhovell** [comment]
cc @MaxGekk I think we need to port this one to 3.3

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you, @hvanhovell . Is it only applicable for master and branch-3.3?
> I think we need to port this one to 3.3

**Comment by viirya** [comment]
s/CODE/CORE/ in the PR title?

**Comment by hvanhovell** [comment]
@dongjoon-hyun the change that introduced this bug is in 3.3 and beyond (it was introduced by yours truly). So we don't need backport it anywhere else.

**Comment by dongjoon-hyun** [comment]
Thank you for the confirmation!

File: core/src/main/scala/org/apache/spark/storage/BlockInfoManager.scala [hunk 1]
```diff
@@ -360,12 +360,17 @@ private[storage] class BlockInfoManager extends Logging {
         info.writerTask = BlockInfo.NO_WRITER
         writeLocksByTask.get(taskAttemptId).remove(blockId)
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/storage/BlockInfoManager.scala:361]: Some other task might be holding the writer lock as well given `releaseAllLocksForTask` - check for that ?

**Review by mridulm — COMMENTED** [review]
The reader side changes look good.

**Comment by dongjoon-hyun** [comment]
I raised the priority of this issue to `Blocker` with target version `3.3.0`.
Although @MaxGekk is in the loop, ping once more.

File: core/src/main/scala/org/apache/spark/storage/BlockInfoManager.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/storage/BlockInfoManager.scala:361]: For this, gentle ping @hvanhovell for his confirmation.

Comment (hvanhovell) [reply core/src/main/scala/org/apache/spark/storage/BlockInfoManager.scala:361]: TBH in that case we are in a world of hurt. You don't know the state of the block (might not be there yet, might be partially materialized, might be ok, or might be deleted) when we release its write lock in `releaseAllLocksForTask`. The problem with releasing is that after the release the block looks fine, and other tasks might try to read it (and fail).

The good news is that the code that takes out the write locks is only in the BlockManager and that it properly cleans up after itself (release locks, remove data if needed). This means we could fix this issue by removing  the release of write locks from `releaseAllLocksForTask`.

**Comment by MaxGekk** [comment]
@hvanhovell @mridulm @dongjoon-hyun @jiangxb1987 This is marked as a blocker for Spark 3.3.0. Any chance that it will be merged to branch-3.3 soon?

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.3!

**Comment by dongjoon-hyun** [comment]
Thank you, @hvanhovell , @cloud-fan , @MaxGekk !

File: core/src/test/scala/org/apache/spark/storage/BlockInfoManagerSuite.scala [hunk 2]
```diff
@@ -360,4 +360,35 @@ class BlockInfoManagerSuite extends SparkFunSuite with BeforeAndAfterEach {
...
+        blockInfoManager.registerTask(task)
+
+        // Acquire read locks
+        (0 to 50).foreach { _ =>
+          assert(blockInfoManager.lockForReading(blockId).isDefined)
+        }
+
+        // Asynchronously release read locks.
+        val futures = (0 to 50).map { _ =>
+          Future(blockInfoManager.unlock(blockId, Option(0L)))
```
Comment (attilapiros) [inline core/src/test/scala/org/apache/spark/storage/BlockInfoManagerSuite.scala:383]: @hvanhovell why `Option(0L)` is passed here and not `Some(task)` as the current task attempt ID was used for the locking too (by the `lockForReading`)? Currently only the first block execution (when task=0) does real testing or am I missing something?

**Pull request closed without merging**